Add Example Workflow for `container-mod` (#70)
ubccr/ccr-examples: containers/1_Advanced/container-mod/README.md

# container-mod Example Workflows

> This directory contains two example workflows ([`personal/`](https://github.com/ubccr/ccr-examples/tree/main/containers/1_Advanced/container-mod/personal) and [`project/`](https://github.com/ubccr/ccr-examples/tree/main/containers/1_Advanced/container-mod/project)) for `container-mod`. The `personal` workflow utilizes a FDS container from DockerHub and runs in default (`personal`) mode. The `project` example is based on a locally stored `.sif` file and an example profile for `container-mod`. More details can be found in [CCR's documentation](https://docs.ccr.buffalo.edu/en/latest/howto/containerization/#container-mod) for `container-mod`.

Reviewer: ```suggestion
This directory contains two example `container-mod` workflows: [`personal/`](https://github.com/ubccr/ccr-examples/tree/main/containers/1_Advanced/container-mod/personal) with files in your home directory and [`project/`](https://github.com/ubccr/ccr-examples/tree/main/containers/1_Advanced/container-mod/project)) with files in (fore example) your projects directory. The `personal` workflow utilizes a FDS container from DockerHub and runs in default (`personal`) mode. The `project` example is based on a locally stored `.sif` file and an example profile for `container-mod`. More details can be found in [CCR's documentation](https://docs.ccr.buffalo.edu/en/latest/howto/containerization/#container-mod) for `container-mod`.
```

## How to use

> 1. Clone the [container-mod repository](https://github.com/TuftsRT/container-mod) into your working directory.

Reviewer: ```suggestion
1. [git clone](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) the [container-mod repository](https://github.com/TuftsRT/container-mod) into your working directory.
```

In this example, we'll use the FDS container image from [DockerHub](https://hub.docker.com/r/satcomx00/fds) and process it with container-mod to generate a ready-to-use environment module. We will use personal mode, where all files are generated in user's `$HOME` directory. More details can be found in [CCR's documentation](https://docs.ccr.buffalo.edu/en/latest/howto/containerization/#container-mod) for `container-mod`.

## How to use

1. Clone the [container-mod repository](https://github.com/TuftsRT/container-mod) into your working directory.

**Important files and their descriptions**
- `container-mod`: Main executable script
- `repos/`: Stores all metadata for all software used by commands such as `module spider`
- `profiles/`: Stores different profiles that define where generated files are placed, instead of the default personal mode (`$HOME`)

**Subcommands used with the `container-mod` script**
- `pull`: Pulls a container image into the desired directory
- `module`: Generates the module file
- `exec`: Generates wrapper scripts (subcommands for the software)
- `pipe`: Important and most useful container-mod command; **runs pull, module and exec in sequence**

> 

Reviewer 2: For each of these sections, if there is a corresponding section in the official repo docs, link to that. For example, add a line here saying "For more information refer to the official [`container-mod` command documentation]((https://github.com/TuftsRT/container-mod#usage).

**Generated directories and files (Personal Mode)**
- `~/container-apps/`: Directory created in personal mode and contains all files and resources used by the generated module
	- `container-apps/images/`: Stores container images pulled from registries such as DockerHub. (Local images stay in their original location)
	- `container-apps/repos/`: Contains metadata for all configured software, used by commands like `module spider`
	- `container-apps/tools/`: Holds generated executables (wrapper scripts) for using specific software programs
- `~/privatemodules/`: Stores generated modulefiles (`.lua`) for Lmod

For detailed information about the topics above and the available options for use, please refer to the [official documentation](https://github.com/TuftsRT/container-mod#usage).

> [!IMPORTANT]
> To use container-mod, first make sure to request an interactive job to utilize Apptainer.

2. Run the `container-mod` script with one of the above Subcommands, followed by the image location. For Example:
```
./container-mod pipe docker://satcomx00/fds:6.7.9
```
(You can also use local images (e.g. from /util/software/containers/); local images stay in their original location) 

> [!NOTE]

> > If no profile is selected, `container-mod` will run in personal mode (all files will be generated in user's `$HOME`). For more information about profiles, please refer to the [official documentation](https://github.com/TuftsRT/container-mod#profiles).

Reviewer 2: If we only support personal mode as in another note, just mention that here and say all files will be in $HOME/subdirectory

3. If there are no software-related files beforehand (module, metadata, executable files), the script will ask the following info only once to create a new entry:

	- Application name, version, description, homepage URL 
	- Available programs: Software commands used to trigger certain behaviors (E.g., fds, abaqus, OpenSees, etc.) 

For an example metadata file, please refer to the official `container-mod` [README](https://github.com/TuftsRT/container-mod#how-it-works).

4. Once the required information is provided, `container-mod` will execute the necessary Apptainer commands on the image and automatically generate the module file if script finishes successfully.

> 5. To use the generated module, you will need to edit your `$MODULEPATH` to include the new path where your module has been generated. You can do this in a few ways:

Reviewer 2: This could be clearer. If we know the file will be in their home directory, specify that here as well.
```suggestion
5. To use the generated module, you will need to edit your `$MODULEPATH` to include `~/container-apps` module has been generated. You can do this in a few ways:
```

	- Use the command (Only eligible for the current session): 
	```

> 	module use ~/privatemodules

Reviewer 2: This was the only method I could get working.

	```

> 	- Edit the `~/.ccr/modulepaths` file to include the full path

Reviewer 2: I'm not sure what the full path is... or at least adding ~/container-apps/ and ~/privatemodules/ to my ~/.ccr/modulepaths and then reloading doesn't show entry in module commands

> 	- Or add the above `module use` command in your `~/.bashrc` so it gets executed at startup (use at your own risk) 

Reviewer 2: I would not do this myself in testing so I would not suggest it to users.
